Sources/TSCBasic/Process.swift

```diff
@@ -531,6 +547,8 @@ public final class Process: ObjectIdentifierProtocol {
             }
         }
       #endif // POSIX implementation
+
+        return stdinStream
```

Comment:
Are we counting on `deinit` closing the stream for the existing clients (or any new clients that ignore the return value)?  Is there any risk in making the existing function return a value that has to be released, for example by turning what would have otherwise been a memory leak (e.g. a captured output value in a block or anything like that) into something that's now holding open the input possibly indefinitely (which depending on the subprocess could mean that it doesn't actually complete)?

Reply:
right, which is why I left the `closeOnDeinit` parameter on the `LocalFileOutputByteStream` initializer, to provide a visual cue that that's the expectation.

This should provide API compatibility with current uses of Process, since it just closes the parent side of the pipe. 

Not sure what you mean by the "risk of returning a value that has to be released" since that seems to be a more cumbersome API with more pitfalls than the current proposal.

Reply:
I should have been clearer.  I was trying to point out what I saw as a potential problem, which is that we're now introducing a return value that might hold the file descriptor open longer than expected.  Is the returned stream guaranteed to be deinited immediately for any existing clients that ignore the return value?

Reply:
not sure about guaranteed, but since there are no references to the `LocalFileOutputByteStream` returned, it does get deinited the moment `launch()` returns, which in turn closes the file, so it's never open longer than required... do you think there are certain scenarios where that won't happen? it seems to me that the guarantee is given by Swift semantics (given that we're not holding a reference to the returned byte stream in the process object itself).

Reply:
That makes sense.  It's probably an unfounded concern, from years of having been burned by Objective-C autorelease pools, which made tying any kind of resource deallocation to object lifetime be very perilous.

So I guess it's not a concern for existing clients, which don't capture the return value, and the only possible thing to watch out for would be a client that captures the return value and then waits for the process to finish, which would be a deadlock if the process waits for the last input from stdin.

For example, something like this:
```
let process = Process(arguments: "shasum")
let writer = process.launch()
writer.write(someBytes)
process.waitForExit()
```
Unless we know for sure that `writer` is deinited before the `waitForExit()`, this would hang.

Reply:
One alternative that I used in a similar implementation (unfortunately not in open source, so I can't point to it) was to have the Process initializer take a block that would be called in the background to write data to the stdin of the process.  The block took a OutputStream as a parameter, and made sure it was closed after the data had been written.  I can see how that can be more constraining for some long-lived processes, though.

Reply:
yeah, that's a good point, it sucks not having an API to close the `OutputByteStream` (other than `let stdinStream: OutputByteStream? = process.launch()` and then `stdinStream = nil`)... 

regarding that example on waiting, it's sort of similar to using a process that expects stdin from a shell script... the only difference is that in the script it's signaled with an `EOF`, and in Swift it would be signalled by nil-ling the stdinStream pointer... unless we added API support for EOF/closing to `OutputByteStream`, not sure what your thoughts are on that

Reply:
Whether or not it's ugly to have an explicit `close()` action in an API like this is a matter of some debate.  My position is that any object that controls an actual resource (other than memory) that has some kind of external semantics, such as a stream, should have some explicit way for a client to say that it's done with that resource.  This is because of the unavoidable side effects of not being able to control the lifetime of the resource.  In other words, it's great that we (for the most part) don't have to think about object lifetimes in Swift, but things like closing a file, stopping listening on a connection, finalizing a database, etc have externally visible semantics and therefore should have explicit control (side note, though: if the user forgets to do it, then deinit should also release the resource automatically).

Interestingly, the file-oriented streams in TSC do have API to close them, we just don't expose it in the generic `OutputStream` that they specialize.  I think we should expose it, to give clients a way to signal intent, even if it's a no-op for some stream implementations (and it's not even always a no-op — for example, even a memory stream could use that as an optimization, by trimming any extraneously allocated memory pages, etc).  Also, having an explicit close action is a way to guarantee that no more information will be written by some other part of the code who might be holding a reference to the stream — that can be important when writing out a file format for example.

So long story short, yes, I do think that we should expose `close` on `OutputByteStream` in general, because in practice it does end up being important, though I know that some people feel it's extraneous.  My own observation is that the more one has been burned by issues around not having explicit control over the end of the stream, the more one tends to want to surface it in the API even though it's slightly unfortunate.  In the end, it ends up being needed.

Reply:
yeah, I agree with you. that said, do you want to add the API to OutputByteStream first and then merge this? or is this PR unblocked from that perspective? I'll have a proposal for the OutputByteStream `close` API in a subsequent PR (and another one to rename OutputByteStream to WritableByteStream)

Reply:
If you need these changes, I wouldn't necessarily block them on reworking the stream.  Maybe you could change the return type to `FileOutputByteStream`, since that's what it always is anyway, which would give callers the `close` function they could use?

Reply:
It would be leaking the implementation a little bit, but having control over the file descriptor is probably worth it.

Reply:
going to do the rename of the stream protocol in https://github.com/apple/swift-tools-support-core/pull/125, and then adding the `close()` API. Seems a cleaner approach to not expose the underlying type, especially since for Windows we're not using `FileOutputByteStream`.

Reply:
https://github.com/apple/swift-tools-support-core/pull/125 for the rename PR

Reply:
Nice, that does seem cleaner.

Reply:
Updated the PR based on this plan, which was executed on #125  and #127